cloudbot/util/decorators.py

```diff
@@ -0,0 +1,25 @@
+from time import time
+
+SPAM_LIST = {}
+
+
+def spam_guard(func):
+    def wrapper(*args, **kwargs):
+        nick = kwargs.get('nick', None)  # pull the nick from the args
+        if nick:  # if there is no nick specified for whatever reason, carry on
+            global SPAM_LIST
+            if SPAM_LIST.get(nick, None):
+                SPAM_LIST[nick][0] = time()
+                if time() - SPAM_LIST[nick][0] < SPAM_LIST[nick][1]:
+                    # notice that they are spamming
+                    SPAM_LIST[nick][1] += 10
+                    return (
+                        "{} attempted to execute a command within their previous" \
+                        " limit. An additional 10 seconds has been added. Current" \
+                        " limit is {} seconds.".format(nick, SPAM_LIST[nick][1])
```

Comment:
Yes: you can choose what functions will be decorated. The problem is that the spam guard does not distinguish between decorated functions in any way; any function call is a possible spam call, so any decorated function call is liable to punish the user for using the system "too quickly". This won't stop dedicated trolls/spammers/people avalanching commands in channel. It's only going to catch people who are a little quicker at using commands than others; e.g. you can't join a session within 10 seconds of finding the one you wanted to join to begin with. That's a completely arbitrary restriction and it aids neither the chat nor the user nor the bot nor its programmers. Even if most people don't have the reaction time or typing speed, it truly solves only the most basic of abuse cases (same user repeating same commands in public chat very quickly) in the most hamfisted of ways. 

There are better ways of solving this problem that will catch more legitimate abuse and still allow people to use the system. Command classes/types (or even just using the passed function's name to distinguish guarded commands) will prevent that same mentioned abuse case without forcing the same restriction on related commands. A more flexible/configurable rate of command use over time will allow people to burst certain commands quickly within reason and still catch actual abuse that goes over the line. An avalanche guard component would count function calls/commands over time not made by a specific user and put a timeout on running that command again in general, when necessary. There are probably more I can't even think of right now. 

All I'm saying is it's more likely to hurt legitimate use than abuse. If that's okay, then godspeed. 

(Although I should also note that it doesn't care if the user is an op or not...)